core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -239,10 +230,24 @@ private[chisel3] trait HasId extends InstanceId {
     refBuilder: String => Arg = Ref(_)
   ): Unit =
     if (_ref.isEmpty) {
-      val candidate_name = _computeName(Some(default).filterNot(_ => errorIfDup)).get
+      val candidate_name = _computeName(Some(default).filterNot(_ => errorIfDup)).getOrElse {
+        throwException(
+          s"Attempted to name a nameless IO port ($this): this is usually caused by instantiating an IO but not assigning it to a val.\nAssign $this to a val, or explicitly call suggestName to seed a unique name"
+        )
+      }
+
+      val sanitized = sanitize(candidate_name)
       val available_name = namespace.name(candidate_name)
-      if (errorIfDup && (available_name != candidate_name)) {
-        Builder.error(s"Cannot have duplicate names $available_name and $candidate_name")(UnlocatableSourceInfo)
+      // Check if characters were stripped from the candidate name and the resulting sanitized name results conflicts with an existing name
+      if (errorIfDup && (candidate_name != sanitized) && (available_name != sanitized)) {
+        Builder.error(
+          s"Attempted to name $this with an unsanitary name '$candidate_name': sanitization results in a duplicated name '$sanitized'. Please seed a more unique name"
+        )(UnlocatableSourceInfo)
+      }
+      if (errorIfDup && (available_name != sanitized)) {
+        Builder.error(
+          s"Attempted to name $this with a duplicated name '$candidate_name'. Use suggestName to seed a unique name"
+        )(UnlocatableSourceInfo)
       }
```

Comment:
Are these errors intended to be exclusive? Because the conditions overlap and these are recoverable errors, if the first 1 executes the 2nd one always will. Also there's a lot of duplication in these conditions, can we make the logic a little tighter?